contract/src/Custody.sol

```diff
@@ -310,54 +293,64 @@ contract Custody is IChannel, IDeposit {
-
-        // Verify that at least one participant signed the state
-        if (candidate.sigs.length == 0) revert InvalidStateSignatures();
+        if (meta.stage == ChannelStatus.DISPUTE || meta.stage == ChannelStatus.FINAL) revert InvalidStatus();
+        if (candidate.intent == StateIntent.FINALIZE) revert InvalidState();
 
-        // Validate version based on channel status
-        if (meta.stage == ChannelStatus.INITIAL && candidate.version != 0) revert InvalidState();
+        StateIntent lastValidStateIntent = meta.lastValidState.intent;
 
-        if (candidate.data.length != 0) {
-            if (candidate.intent == StateIntent.INITIALIZE) {
-                // TODO:
-            } else if (candidate.intent == StateIntent.RESIZE) {
-                uint256 deposited = meta.expectedDeposits[CLIENT_IDX].amount + meta.expectedDeposits[SERVER_IDX].amount;
-                uint256 expected = candidate.allocations[CLIENT_IDX].amount + candidate.allocations[SERVER_IDX].amount;
-                if (deposited != expected) {
-                    revert InvalidAllocations();
-                }
+        if (meta.stage == ChannelStatus.INITIAL) {
+            // main goal: verify Candidate == LastValidState, close channel
+            if (!Utils.statesAreEqual(candidate, meta.lastValidState)) {
+                revert InvalidState();
             }
-        }
 
-        if (
-            candidate.data.length == 0
-                || (candidate.intent != StateIntent.INITIALIZE && candidate.intent != StateIntent.RESIZE)
-        ) {
-            // if no state data or intent is not INITIALIZE or RESIZE, assume this is a normal state
+            _closeEffectsAndInteractions(channelId, candidate.allocations);
 
-            // Verify the state is valid according to the adjudicator
-            if (!IAdjudicator(meta.chan.adjudicator).adjudicate(meta.chan, candidate, proofs)) revert InvalidState();
+            emit Challenged(channelId, block.timestamp);
+            emit Closed(channelId, candidate);
+            return;
+        }
 
-            // Reject states with equal version
-            if (candidate.version == meta.lastValidState.version) {
-                // Explicitly check for equal versions and reject them
-                revert InvalidState();
+        // meta.stage == ChannelStatus.ACTIVE
+        // main goal: verify Candidate is valid and >= LastValidState (in RESIZE case states should be equal)
+        if (lastValidStateIntent == StateIntent.INITIALIZE) {
+            if (candidate.intent == StateIntent.INITIALIZE) {
+                if (!Utils.statesAreEqual(candidate, meta.lastValidState)) revert InvalidState();
+            } else {
+                if (!_isMoreRecent(meta.chan.adjudicator, candidate, meta.lastValidState)) revert InvalidState();
+                if (!IAdjudicator(meta.chan.adjudicator).adjudicate(meta.chan, candidate, proofs)) {
+                    revert InvalidState();
+                }
             }
```

Comment:
Is it possible to send INITILIZE state in non-INITIAL channel, that would be more recent than the previous (presumably the first one)?

Reply:
Also, what's the point in not reverting states that are equal - it's just brning gas without changing SC memory state?

Reply:
> Is it possible to send INITILIZE state in non-INITIAL channel, that would be more recent than the previous (presumably the first one)?

If Adjudicator is implemented correctly, than it should not be possible to do so. However, we would like to protect from such case on the protocol level, and thus such checks.
Moreover, `StateIntent` and `ChannelStatus` enums logic partially overlap one another, which causes such odd checks. This should be fixed in future with combining these enums and / or making some states implicit, i.e. defined by some channel storage variables.

Reply:
> Also, what's the point in not reverting states that are equal - it's just brning gas without changing SC memory state?

Channel still records a challenge if states are equal. The logic is different because if challenged and last valid state are equal, we don't need to call `adjudicate(candidate)`, and thus we save gas.